### PR TITLE
Django command for generating waveforms

### LINKED ISSUE #529: Audio waveform cache-warming Django command
## Description

Related to #510

We now have a table where audio waveforms are cached, but this table is not yet fully populated. We have plans to move this processing upstream into the catalog (https://github.com/WordPress/openverse-catalog/issues/363). 

For the time being, we should create a Django command that can be run with `python manage.py <command>` which will iterate through all audio records and generate waveforms for them. Here is an outline of the necessary code:

```python
audios = Audio.objects.all()
with tqdm(total=audios.count()) as progress:
    for audio in audios:
        audio.get_or_create_waveform()
        progress.update(1)
```

The above code example uses `tqdm`, which I don't believe is a dependency we have right now. We could add it for this use and remove it after, since the progress bar and estimated time to completion are very useful. 

We have ~14k records, and a test run on staging showed that this could take anywhere from 12-48 hours.
image (3)

## Implementation

- [ ] 🙋 I would be interested in implementing this feature.

### PR DESCRIPTION
**Pull request by AetherUnbound**
## Fixes

Fixes #529 by @AetherUnbound

## Description

This PR adds a Django command, `generatewaveforms`, for generating waveforms of all audio records in order to populate the waveform cache.

I've added a new dependency, [`django-tqdm`](https://pypi.org/project/django-tqdm/), to make the output easier to read. We can remove this dependency once the API no longer generates waveforms.

**Command list**

```bash
root@c219265a4b39:/api# python manage.py --help

Type 'manage.py help <subcommand>' for help on a specific subcommand.

Available subcommands:

[auth]
    changepassword
    createsuperuser

[catalog]
    generatewaveforms

[contenttypes]
    remove_stale_contenttypes

...
```

**Help text**
```bash
root@c219265a4b39:/api# python manage.py generatewaveforms --help
usage: manage.py generatewaveforms [-h] [--version] [-v {0,1,2,3}] [--settings SETTINGS] [--pythonpath PYTHONPATH] [--traceback] [--no-color] [--force-color] [--skip-checks]

Generates waveforms for all audio records to populate the cache

options:
  -h, --help            show this help message and exit
  --version             show program's version number and exit
  -v {0,1,2,3}, --verbosity {0,1,2,3}
                        Verbosity level; 0=minimal output, 1=normal output, 2=verbose output, 3=very verbose output
  --settings SETTINGS   The Python path to a settings module, e.g. "myproject.settings.main". If this isn't provided, the DJANGO_SETTINGS_MODULE environment variable will be used.
  --pythonpath PYTHONPATH
                        A directory to add to the Python path, e.g. "/home/djangoprojects/myproject".
  --traceback           Raise on CommandError exceptions
  --no-color            Don't colorize the command output.
  --force-color         Force colorization of the command output.
  --skip-checks         Skip system checks.
```

**Sample output**
```bash
root@c219265a4b39:/api# python manage.py generatewaveforms
Generating waveforms for 5,000 records
  0%|                                                                                                                            | 1/5000 [00:00<16:19,  5.10it/s]
  Unable to process 67acc776-838c-4242-bb99-be9e0be061b6: Can't generate json format output from Unknown file format
  1%|██▍                                                                                                                        | 69/5000 [01:04<45:46,  1.80it/s]
```

## Testing Instructions

1. `just recreate`
1. `just init`
1. `docker-compose exec web bash`
1. `python manage.py generatewaveforms`

## Checklist

- [x] My pull request has a descriptive title (not a vague title like `Update index.md`).
- [x] My pull request targets the *default* branch of the repository (`main`) or a parent feature branch.
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [ ] I added or updated tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no visible errors.

[best_practices]:https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

**Review by dhruvkb — APPROVED** [review]

**Review by sarayourfriend — COMMENTED** [review]
Tested it locally and it worked! I ran the command and then `just dj shell` and confirmed the processed waveforms were in the audio_add_on table.

I've got a couple questions/gut-checks I'd like to verify before approving. Unit tests would also be nice to have for this, especially if my suggestions end up getting implemented (in particular the start/stop/resume aspect) and the complexity of it increases.

File: api/catalog/management/commands/generatewaveforms.py
Comment (AetherUnbound) [reply api/catalog/management/commands/generatewaveforms.py:16]: Ah drat, I naively assumed that iterating over them in this manner would automatically paginate a cursor on the database. According to the docs[^1], `all()` evaluates the entire `QuerySet`. I think we can use a `Paginator` to improve that.

> Would it be worth excluding any rows that already have corresponding rows in the add_on table?

Absolutely, great call!
[^1]: https://docs.djangoproject.com/en/dev/ref/models/querysets/#when-querysets-are-evaluated

File: api/catalog/management/commands/generatewaveforms.py
Comment (AetherUnbound) [reply api/catalog/management/commands/generatewaveforms.py:24]: I would be okay with adding a delay at this step :slightly_smiling_face: 0.25 seconds is a small fraction of the time it takes to actually retrieve + compute the waveform, so that seems fine.

**Comment by AetherUnbound** [comment]
> Unit tests would also be nice to have for this, especially if my suggestions end up getting implemented (in particular the start/stop/resume aspect) and the complexity of it increases.

Agreed, I'll try to start on those if I have time before I'm out. Initially this was really straightforward, but then I added error handling, and with the other suggestions it definitely makes sense to do!

File: api/catalog/management/commands/generatewaveforms.py
Comment (sarayourfriend) [reply api/catalog/management/commands/generatewaveforms.py:24]: I was noticing that as I ran it locally. I _almost_ suggested doing something like a speed _limit_ (1 row per second but include the time spent processing the previous row in the "wait" time between each row... or something?) but decided that was probably too complicated/overkill (though if there's some kind of utility for already doing that easily it would be sweet to use it).

**Comment by sarayourfriend** [comment]
@AetherUnbound If you're okay with handing this off and run out of time, I'm happy to pick this up wherever you leave off, no sweat, just let me know.

**Comment by AetherUnbound** [comment]
@sarayourfriend I don't think I'm going to end up getting to this before I'm out - please feel free to take it over!

**Comment by sarayourfriend** [comment]
@AetherUnbound Sure thing!

**Comment by sarayourfriend** [comment]
@obulat @dhruvkb this should be ready for review now after adding pagination and unit tests

**Review by dhruvkb — APPROVED** [review]
This looks good to me. A few lint fixes pointed out by the check and we could merge this!

**Comment by zackkrida** [comment]
@sarayourfriend Seeing three remaining lint errors in the [latest run](https://github.com/WordPress/openverse-api/runs/5394971440?check_suite_focus=true):

```shell
api/test/unit/management/commands/generatewaveforms_test.py:53:5: F841 local variable 'waveforms' is assigned to but never used
api/test/factory/models/media.py:6:1: F401 'catalog.api.models.media.AbstractMedia' imported but unused
api/test/factory/models/media.py:21:58: E741 ambiguous variable name 'l'
```

**Comment by sarayourfriend** [comment]
Thanks @zackkrida, fixed and pushed. Tho now there are some conflicts I need to resolve :+1:

File: api/catalog/api/models/audio.py
Comment (obulat) [inline api/catalog/api/models/audio.py:195]: Left-overs from merge. They cause linting failure in CI, too.

**Review by obulat — COMMENTED** [review]
I first tried it out with the `get_waveform` leftover from the merge, and then with the `get_waveform removed. I guess the latter is the correct way.

It works, but it seems like it's stopping every once in a while. I think that it could be swallowing some errors. Below is an error I got when I was running it at first. The process stalled, and I pressed CTRL+C, and this was the error I saw in the console:
```
Unable to process d495818e-b569-45a6-8e16-64f37f672225: HTTPSConnectionPool(host='mp3d.jamendo.com', port=443): Max retries exceeded with url: /download/track/1204669/mp32 (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0xffff948bf460>: Failed to establish a new connection: [Errno 111] Connection refused'))
  0%|                                       | 6/5000 [01:19<25:25:55, 18.33s/it]^CKeyboardInterrupt
2022-03-03T07:54:38Z
Unable to process 2470ec1f-0c29-4063-98d9-8c4da09c3572:
```
The process went on after CTRL+C, only to stop again later.

This error seems to be different from `  1%|▋                                                                                 | 39/5000 [00:55<49:37,  1.67it/s]Unable to process 67acc776-838c-4242-bb99-be9e0be061b6: Can't generate json format output from Unknown file format`, too.

1. Why is the waveform generation not resuming after an error?
2. I think we should record the files that cause the error to check them and possibly remove from the catalog if they are invalid.

File: api/catalog/management/commands/generatewaveforms.py
Comment (obulat) [inline api/catalog/management/commands/generatewaveforms.py:32]: I think it should depend on the size of the audio file. Many Freesound and Wikimedia files are very small. I sometimes see 2-3 it/s in the console, and the maximum is 11 it/s.

Comment (sarayourfriend) [reply api/catalog/management/commands/generatewaveforms.py:32]: Right. I think we probably should use the `limit` library anyway just in case.

File: api/catalog/management/commands/generatewaveforms.py
Comment (AetherUnbound) [inline api/catalog/management/commands/generatewaveforms.py:42]: Just a thought, the fact that these both have `limit` in them is kind of confusing. Maybe we could do:
* `--unlimited` -> `--no-rate-limit`
* `--limit` -> `--max-records`

What do you think?

File: api/catalog/management/commands/generatewaveforms.py
Comment (AetherUnbound) [inline api/catalog/management/commands/generatewaveforms.py:95]: Are we taking this approach because `Audio` and `AudioAddOn` don't have a foreign key relationship?

File: api/test/factory/models/audio.py
Comment (AetherUnbound) [inline api/test/factory/models/audio.py:8]: Super neat!

File: api/test/run_test.sh
Comment (AetherUnbound) [inline api/test/run_test.sh:22]: I know we have another ticket for this - should we hold off on this change for now?

File: api/test/unit/management/commands/generatewaveforms_test.py
Comment (AetherUnbound) [inline api/test/unit/management/commands/generatewaveforms_test.py:44]: I love this approach!

File: api/test/unit/management/commands/generatewaveforms_test.py
Comment (AetherUnbound) [inline api/test/unit/management/commands/generatewaveforms_test.py:89]: I know the format isn't great as a result of this change, but if these comments are on the line above, `black` won't try and split L87-L89 over multiple lines

**Review by AetherUnbound — COMMENTED** [review]
I can't approve my own PR :sweat_smile: but here's a few comments, nothing to stop a merge though!

File: api/catalog/management/commands/generatewaveforms.py
Comment (sarayourfriend) [reply api/catalog/management/commands/generatewaveforms.py:42]: Yes, I like these better. I hated the ones I chose!

File: api/catalog/management/commands/generatewaveforms.py
Comment (sarayourfriend) [reply api/catalog/management/commands/generatewaveforms.py:95]: By "this approach" what do you mean? The inner select on the add ons table?

The tables don't have FK relationship though, you're right, at least not one the DB knows about.

File: api/test/run_test.sh
Comment (sarayourfriend) [reply api/test/run_test.sh:22]: Oh sure. I think I did this in rage at not being able to debug the tests, didn't mean to commit it though.

File: api/test/unit/management/commands/generatewaveforms_test.py
Comment (sarayourfriend) [reply api/test/unit/management/commands/generatewaveforms_test.py]: Done in 8fb6223843372f6714aaea31dc1ea0545751c9f7

File: api/catalog/management/commands/generatewaveforms.py
Comment (sarayourfriend) [reply api/catalog/management/commands/generatewaveforms.py]: Done in e75964f68a12029b70c8c950945dcc0a9f907a95

File: api/test/run_test.sh
Comment (sarayourfriend) [reply api/test/run_test.sh:22]: Put back in 753182c436281804b2e8508210da8c01f1ac32fc

File: api/catalog/management/commands/generatewaveforms.py
Comment (AetherUnbound) [reply api/catalog/management/commands/generatewaveforms.py:95]: Oh sorry, I meant that I could have sworn we could do something like
```
Audio.objects.filter(audioaddon__waveform_peaks__isnull=True)
```
if these were foreign keys. They're not, so I don't think Django would be able to interpret that relationship appropriately.

_And sorry for the lack of context, I had that initially but second guessed myself and deleted the entire comment :woman_facepalming: _

**Comment by zackkrida** [comment]
@AetherUnbound I think your symbolic approval here would be fine, once you think it's good to go I can merge it, if no one else reviews it by then.

**Comment by stacimc** [comment]
This looks great! Manually testing with interrupts, with the rate limit enabled/disabled, and setting a limit to let the batch complete all seem to be working great 🎉 

Other than merge conflicts, I think the [`test_audio_success_examples`](https://github.com/WordPress/openverse-api/blob/main/api/test/examples_test.py#L25) test may also need to be updated. When I run the api-tests locally I get something like:
```
Differing items:
E         {'peaks': [0.61275, 0.19593, 0.74023, 0.45534, 0.52315, 0.40059, ...]} != {'peaks': []}
```
For the audio detail & search endpoints.

**Comment by AetherUnbound** [comment]
I'm also getting the error Staci is getting, as well as:

```
in_val = '\n# Search for music titled "Wish You Were Here" by The.madpix.project\ncurl  "http://localhost:8000/v1/audio/?title=Wish%20You%20Were%20Here&creator=The.madpix.project"\n'
out_val = {'application/json': {'page': 1, 'page_count': 0, 'page_size': 20, 'result_count': 1, ...}}

    @pytest.mark.parametrize("in_val, out_val", list(audio_mappings.items()))
    def test_audio_success_examples(in_val, out_val):
        res = execute_request(in_val)
>       assert res == out_val["application/json"]
E       AssertionError: assert {'page': 1, '...ount': 1, ...} == {'page': 1, '...ount': 1, ...}
E         Omitting 4 identical items, use -vv to show
E         Differing items:
E         {'results': [{'category': 'music', 'creator': 'The.madpix.project', 'creator_url': 'https://www.jamendo.com/artist/441585/the.madpix.project', 'detail_url': 'http://localhost:8000/v1/audio/8624ba61-57f1-4f98-8a85-ece206c319cf/', ...}]} != {'results': [{'category': 'music', 'creator': 'The.madpix.project', 'creator_url': 'https://www.jamendo.com/artist/441585/the.madpix.project', 'detail_url': 'http://localhost:8000/v1/audio/8624ba61-57f1-4f98-8a85-ece206c319cf/', ...}]}
E         Use -v to get the full diff
```

File: api/catalog/management/commands/generatewaveforms.py
Comment (sarayourfriend) [reply api/catalog/management/commands/generatewaveforms.py:95]: Ohh yes no that's not possible unfortunately because without the explicit relationship field Django won't add the reverse field relationships you'd need to make that JOIN.

**Comment by sarayourfriend** [comment]
@stacimc @AetherUnbound does that happen on a fresh version of the API or only with existing data after running `generatewaveforms`? 

If I do `just recreate && just init && just api-tests` they all pass. But if I run `just dj "generatewaveforms --max_records=3"` and then run `just api-tests` again they fail.

I think this is because our API integration tests rely on the existing database rather than (what would be ideal) building their own data that then gets torn down after each test iteration.

Is fixing that in scope for this issue? I could probably find a workaround but it would likely cheapen the tests :confused: The easiest thing I can think would be to just delete the `peaks` property from both sides of the comparison.

**Review by obulat — APPROVED** [review]
I tried generating the waveforms and running together with the frontend locally, and I got myself some instant waveforms! :)
The problem is that there is a mismatch between what the front end assumes and what the API returns when no waveforms are available. If I understand correctly, the API returns an empty list `[]` if the waveform is not available.
The frontend expects either an array with at least 1 item where all values are between 0 and 1, or a falsy value (then it generates random peaks). 
When the frontend gets an empty list for peaks, `[]`, it does not show any waveform,  the validator throws an error, and there are lots of `Nan` errors when it is trying to calculate the waveform peaks.
I am not sure whether it should be fixed in this PR, or a separate one, though, so I'm approving :)

**Comment by AetherUnbound** [comment]
Ah, my bad! I didn't run `recreate` before testing. I'm doing so now, as soon as that finishes OK I'll fix the Pipfile.lock conflicts and merge :smile:

**Comment by sarayourfriend** [comment]
@obulat that's a bug introduced by the previous PR that actually added the peaks data to the API. Let's fix it in a separate issue.

**Review by stacimc — APPROVED** [review]
> I think this is because our API integration tests rely on the existing database rather than (what would be ideal) building their own data that then gets torn down after each test iteration.

My bad, I didn't look closely enough at the tests. They _are_ passing after recreating/reinitializing everything. That's definitely not ideal but absolutely a separate issue and shouldn't be fixed here, agreed!

**Pull request merged**
